dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClientMultipleDMLOperationTest.java

```diff
@@ -0,0 +1,1006 @@
+                                    return null;
+                                });
+                    }
+                }
+                break;
+        }
+        Timer timer = new Timer(20);
+        while (this.totalDMLOperation.get() < maxIteration) {
+            try {
+                if (!this.failedMessage.isEmpty())
+                    fail(String.format("Failed to %s data with error: %s", dmlOperation, this.failedMessage));
+                if (timer.expired()) {
+                    fail(String.format("Only %d requests completed after %d sec.",
+                            this.totalDMLOperation.get(), timer.getTimeout()));
+                }
+                Thread.sleep(100);
+            }  catch (InterruptedException e) {
+                break;
+            }
+        }
+    }
+
+    static class Timer {
+        private final long endTime;
+        private final int timeOut;
+
+        public Timer(int timeOut) {
+            this.timeOut = timeOut;
+            this.endTime = System.currentTimeMillis() + 1_000L * timeOut;
+        }
+
+        public boolean expired() {
+            return System.currentTimeMillis() >= this.endTime;
+        }
+
+        int getTimeout() {
+            return this.timeOut;
+        }
+    }
+
+    static class MockConnectionPool implements io.helidon.dbclient.jdbc.ConnectionPool {
+        int maxPoolCount = 10;
+        List <Connection> connectionPool = Collections.synchronizedList(new ArrayList<>(maxPoolCount));
+        List <Connection> usedConnections = Collections.synchronizedList(new ArrayList<>(maxPoolCount));
+
+        int actualConnection = 0;
+
+        @Override
+        public Connection connection() {
+            Connection conn;
+            this.actualConnection++;
+            // get connection from the pool if it is not empty,
+            if (!connectionPool.isEmpty()) {
+                conn = this.connectionPool.remove(0);
+            } else {
+                Timer timer = new Timer(10);
+                // if usedConnections reach maxPoolCount, wait for 5 minutes until it recedes. Otherwise throw an exception
+                while (this.usedConnections.size() >= maxPoolCount) {
+                    try {
+                        if (timer.expired()) {
+                            throw new DbClientException(
+                                    String.format("Unable to acquire connection after %d sec", timer.getTimeout()));
+                        }
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {
+                        throw new DbClientException(
+                                String.format("Unable to acquire connection due to InterruptedException: %s", e.getMessage()));
+                    }
+                }
+                // Create a mocked up Connection that returns a PreparedStatement
+                conn = new Connection() {
```

Comment:
This seems like using `Mockito` would be good, as you only need to implement a few methods. It also allows you to validate what was called.

Reply:
I initially wanted to use Mockito, but the few implemented methods need to do more than just return a value. For example, Connection.close() would be doing this:

```
                    @Override
                    public void close() throws SQLException {
                        if (usedConnections.remove(this)) {
                            connectionPool.add(this);
                        }
                        LOGGER.fine(() ->
                                String.format("After Connection.close() metrics: connectionPool=%d, usedConnections=%d",
                                connectionPool.size(), usedConnections.size()));
                    }
```

I thought Mockito can't do such? But I will look at it some more to see if it is possible.